FrameGraph: add support for SSR

Popov72 opened the pull request:
Sorry, big PR!

bjsplat commented:
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

sebavan commented on packages/dev/core/src/Engines/nativeEngine.ts:
```diff
@@ -945,6 +945,10 @@ export class NativeEngine extends Engine {
         this._commandBufferEncoder.finishEncodingCommand();
     }
 
+    public override setStateCullFaceType(_cullBackFaces?: boolean, _force?: boolean): void {
```
Not sure if we need a new function or we could do it within the state ? knowing that otherwise it also needs to be in native ?

Popov72 replied:
I needed a separate function for the geometry renderer (https://github.com/BabylonJS/Babylon.js/pull/16262/files#diff-fc2462ee78aa57d4d2e939a53a4dda62237a6f11a0f2ca0343df95c3bfce77d6R291-R293), because I must be able to reverse this setting on user request (we need to generate a depth buffer from back faces when we use the automatic thickness computation in SSR).

bjsplat commented:
Snapshot stored with reference name:
refs/pull/16262/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16262/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16262/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/16262/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/16262/merge
https://gui.babylonjs.com/?snapshot=refs/pull/16262/merge
https://nme.babylonjs.com/?snapshot=refs/pull/16262/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/16262/merge#BCU1XR#0

bjsplat commented:
You have made possible changes to one or more graph tools. To test the snapshots:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/nodeEditor/
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/nodeGeometryEditor/
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/nodeRenderGraphEditor/
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/guiEditor/

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16262/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16262/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
You have made possible changes to one or more graph tools. To test the snapshots:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/nodeEditor/
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/nodeGeometryEditor/
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/nodeRenderGraphEditor/
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/guiEditor/

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16262/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16262/merge/testResults/webgpuplaywright/index.html

Popov72 commented:
The "Custom render target" test was quite flaky. I think it's because we didn't check `scene.customRenderTargets` for readiness, so I've added the check and removed the "renderCount: 20" entry of the json file.

bjsplat commented:
You have made possible changes to one or more graph tools. To test the snapshots:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/nodeEditor/
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/nodeGeometryEditor/
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/nodeRenderGraphEditor/
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/guiEditor/

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16262/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16262/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
You have made possible changes to one or more graph tools. To test the snapshots:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/nodeEditor/
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/nodeGeometryEditor/
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/nodeRenderGraphEditor/
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/TOOLS/refs/pull/16262/merge/guiEditor/

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16262/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/16262/merge/testResults/webgpuplaywright/index.html
